src/amplify-migration/jsDomUtils.ts

```diff
@@ -0,0 +1,186 @@
+import { JSDOM } from "jsdom";
+import assert from "assert";
+import fs from "fs";
+import os from "os";
+import path from "path";
+import { errorMessage } from "./errorMessage";
+import { REPOS_WITH_ERRORS } from "./constants";
+
+type TagAttribute<T extends "a" | "img"> = T extends "a"
+  ? { tagName: "a"; attribute: "href" }
+  : { tagName: "img"; attribute: "src" };
+export async function modifyTagAttribute({
+  dom,
+  tagAttribute: { tagName, attribute },
+  changedPermalinks,
+  fileContent,
+  setOfAllDocumentsPath,
+  currentRepoName,
+}: {
+  dom: JSDOM;
+  tagAttribute: TagAttribute<"a" | "img">;
+  changedPermalinks: { [oldPermalink: string]: string };
+  fileContent:string
+  setOfAllDocumentsPath: Set<string>;
+  normalisedUrls: Set<string>;
+  currentRepoName: string;
+}): Promise<{
+  fileContent: string;
+  dom: JSDOM;
+  changedPermalinks: { [oldPermalink: string]: string };
+}> {
+
+  for (const tag of dom.window.document.querySelectorAll(tagName)) {
+    // replace permalinks with lowercase and in changedPermalinks
+    if (attribute === "href") {
+      /**
+       * Ideally this code should be const a = tag as HTMLAnchorElement;
+       * However, HTMLAnchorElement is a built-in interface in the browser's DOM API,
+       * which means that it's only available in a browser environment.
+       * In development, hovering over tag will show the inferred type as HTMLAnchorElement
+       */
+      const a = tag as any;
+
+      let rawPermalink = getRawPermalink(a[attribute]);
+
+      if (changedPermalinks[rawPermalink]) {
+        a[attribute] = a.href.replace(
+          rawPermalink,
+          changedPermalinks[rawPermalink]
+        );
+        fileContent = fileContent.replace(
```

Comment:
@harishv7 eg. here, we find the a.href using JSDOM, but we do a literal string replace here, bypassing the HTML normalisation